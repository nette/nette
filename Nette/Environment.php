<?php

/**
 * Nette Framework
 *
 * Copyright (c) 2004, 2009 David Grudl (http://davidgrudl.com)
 *
 * This source file is subject to the "Nette license" that is bundled
 * with this package in the file license.txt.
 *
 * For more information please see http://nettephp.com
 *
 * @copyright  Copyright (c) 2004, 2009 David Grudl
 * @license    http://nettephp.com/license  Nette license
 * @link       http://nettephp.com
 * @category   Nette
 * @package    Nette
 * @version    $Id$
 */

/*namespace Nette;*/



/**
 * Nette environment and configuration.
 *
 * @author     David Grudl
 * @copyright  Copyright (c) 2004, 2009 David Grudl
 * @package    Nette
 */
final class Environment
{
	/**#@+ environment name */
	const DEVELOPMENT = 'development';
	const PRODUCTION = 'production';
	const CONSOLE = 'console';
	const LAB = 'lab';
	/**#@-*/

	/**#@+ mode name */
	const DEBUG = 'debug';
	const PERFORMANCE = 'performance';
	/**#@-*/

	/** @var Configurator */
	private static $configurator;

	/** @var string  the mode of current application */
	private static $mode = array();

	/** @var \ArrayObject */
	private static $config;

	/** @var IServiceLocator */
	private static $serviceLocator;

	/** @var array */
	private static $vars = array(
		'encoding' => array('UTF-8', FALSE),
		'lang' => array('en', FALSE),
		'cacheBase' => array('%tempDir%/cache-', TRUE),
		'tempDir' => array('%appDir%/temp', TRUE),
		'logDir' => array('%appDir%/log', TRUE),
		'templatesDir' => array('%appDir%/templates', TRUE),
		'presentersDir' => array('%appDir%/presenters', TRUE),
		'componentsDir' => array('%appDir%/components', TRUE),
		'modelsDir' => array('%appDir%/models', TRUE),
	);



	/**
	 * Static class - cannot be instantiated.
	 */
	final public function __construct()
	{
		throw new /*\*/LogicException("Cannot instantiate static class " . get_class($this));
	}



	/**
	 * Sets "class behind Environment" configurator.
	 * @param  Configurator
	 * @return void
	 */
	public static function setConfigurator(Configurator $configurator)
	{
		self::$configurator = $configurator;
	}



	/**
	 * Gets "class behind Environment" configurator.
	 * @return Configurator
	 */
	public static function getConfigurator()
	{
		if (self::$configurator === NULL) {
			self::$configurator = new Configurator;
		}
		return self::$configurator;
	}



	/********************* environment name and modes ****************d*g**/



	/**
	 * Sets the current environment name.
	 * @param  string
	 * @return void
	 * @throws \InvalidStateException
	 */
	public static function setName($name)
	{
		if (!isset(self::$vars['environment'])) {
			self::setVariable('environment', $name, FALSE);

		} else {
			throw new /*\*/InvalidStateException('Environment name has been already set.');
		}
	}



	/**
	 * Returns the current environment name.
	 * @return string
	 */
	public static function getName()
	{
		$name = self::getVariable('environment');
		if ($name === NULL) {
			$name = self::getConfigurator()->detect('environment');
			self::setVariable('environment', $name, FALSE);
		}
		return $name;
	}



	/**
	 * Sets the mode.
	 *
	 * @param  string mode identifier
	 * @param  bool   set or unser
	 * @return void
	 */
	public static function setMode($mode, $flag = TRUE)
	{
		if ($mode === 'live') {
			trigger_error("Environment mode 'live' is deprecated; use 'production' instead.", E_USER_WARNING);
			$mode = 'production';
		}
		self::$mode[$mode] = (bool) $flag;
	}



	/**
	 * Returns the mode.
	 *
	 * @param  string mode identifier
	 * @return bool
	 */
	public static function getMode($mode)
	{
		if ($mode === 'live') {
			trigger_error("Environment mode 'live' is deprecated; use 'production' instead.", E_USER_WARNING);
			$mode = 'production';
		}
		if (isset(self::$mode[$mode])) {
			return self::$mode[$mode];

		} else {
			return self::$mode[$mode] = self::getConfigurator()->detect($mode);
		}
	}



	/**
	 * Detects console (non-HTTP) mode.
	 * @return bool
	 */
	public static function isConsole()
	{
		return self::getMode('console');
	}



	/**
	 * Determines whether a server is running in production mode.
	 * @return bool
	 */
	public static function isProduction()
	{
		return self::getMode('production');
	}



	/**
	 * @deprecated {@link Environment::isProduction()}
	 */
	public static function isLive()
	{
		trigger_error('Environment::isLive() is deprecated; use Environment::isProduction() instead.', /**/E_USER_WARNING/**//*E_USER_DEPRECATED*/);
		return self::getMode('production');
	}



	/**
	 * Determines whether the debugger is active.
	 * @return bool
	 */
	public static function isDebugging()
	{
		return self::getMode('debug');
	}



	/********************* environment variables ****************d*g**/



	/**
	 * Sets the environment variable.
	 * @param  string
	 * @param  mixed
	 * @param  bool
	 * @return void
	 */
	public static function setVariable($name, $value, $expand = TRUE)
	{
		if (!is_string($value)) {
			$expand = FALSE;
		}
		self::$vars[$name] = array($value, (bool) $expand);
	}



	/**
	 * Returns the value of an environment variable or $default if there is no element set.
	 * @param  string
	 * @param  mixed  default value to use if key not found
	 * @return mixed
	 * @throws \InvalidStateException
	 */
	public static function getVariable($name, $default = NULL)
	{
		if (isset(self::$vars[$name])) {
			list($var, $exp) = self::$vars[$name];
			if ($exp) {
				$var = self::expand($var);
				self::$vars[$name] = array($var, FALSE);
			}
			return $var;

		} else {
			// convert from camelCaps (or PascalCaps) to ALL_CAPS
			$const = strtoupper(preg_replace('#(.)([A-Z]+)#', '$1_$2', $name));
			$list = get_defined_constants(TRUE);
			if (isset($list['user'][$const])) {
				self::$vars[$name] = array($list['user'][$const], FALSE);
				return $list['user'][$const];

			} else {
				return $default;
			}
		}
	}



	/**
	 * Define one or more variables as constants.
	 * @param  string|array
	 * @return void
	 */
	public static function exportConstant($names)
	{
		if (!is_array($names)) {
			$names = func_get_args();
		}

		foreach ($names as $name) {
			$const = strtoupper(preg_replace('#(.)([A-Z]+)#', '$1_$2', $name));
			define($const, self::getVariable($name));
		}
	}



	/**
	 * Returns expanded variable.
	 * @param  string
	 * @return string
	 * @throws \InvalidStateException
	 */
	public static function expand($var)
	{
		if (is_string($var) && strpos($var, '%') !== FALSE) {
			return @preg_replace_callback('#%([a-z0-9_-]*)%#i', array(__CLASS__, 'expandCb'), $var); // intentionally @ due PHP bug #39257
		}
		return $var;
	}



	/**
	 * @see Environment::expand()
	 * @param  array
	 * @return string
	 */
	private static function expandCb($m)
	{
		list(, $var) = $m;
		if ($var === '') return '%';

		static $livelock;
		if (isset($livelock[$var])) {
			throw new /*\*/InvalidStateException("Circular reference detected for variables: "
				. implode(', ', array_keys($livelock)) . ".");
		}

		try {
			$livelock[$var] = TRUE;
			$val = self::getVariable($var);
			unset($livelock[$var]);
		} catch (/*\*/Exception $e) {
			$livelock = array();
			throw $e;
		}

		if ($val === NULL) {
			throw new /*\*/InvalidStateException("Unknown environment variable '$var'.");

		} elseif (!is_scalar($val)) {
			throw new /*\*/InvalidStateException("Environment variable '$var' is not scalar.");
		}

		return $val;
	}



	/********************* service locator ****************d*g**/



	/**
	 * Get initial instance of service locator.
	 * @return IServiceLocator
	 */
	public static function getServiceLocator()
	{
		if (self::$serviceLocator === NULL) {
			self::$serviceLocator = self::getConfigurator()->createServiceLocator();
		}
		return self::$serviceLocator;
	}



	/**
	 * Gets the service object of the specified type.
	 * @param  string service name
	 * @param  bool   throw exception if service doesn't exist?
	 * @return mixed
	 */
	static public function getService($name, $need = TRUE)
	{
		return self::getServiceLocator()->getService($name, $need);
	}



	/**
	 * @return Nette\Web\IHttpRequest
	 */
	public static function getHttpRequest()
	{
		return self::getServiceLocator()->getService('Nette\Web\IHttpRequest');
	}



	/**
	 * @return Nette\Web\IHttpResponse
	 */
	public static function getHttpResponse()
	{
		return self::getServiceLocator()->getService('Nette\Web\IHttpResponse');
	}



	/**
	 * @return Nette\Application\Application
	 */
	public static function getApplication()
	{
		return self::getServiceLocator()->getService('Nette\Application\Application');
	}



	/**
	 * @return Nette\Web\IUser
	 */
	public static function getUser()
	{
		return self::getServiceLocator()->getService('Nette\Web\IUser');
	}



	/********************* service factories ****************d*g**/



	/**
	 * @param  string
	 * @return Nette\Caching\Cache
	 */
	public static function getCache($namespace = '')
	{
		return new /*Nette\Caching\*/Cache(
			self::getService('Nette\Caching\ICacheStorage'),
			$namespace
		);
	}



	/**
	 * Returns instance of session or session namespace.
	 * @param  string
	 * @return Nette\Web\Session|Nette\Web\Session
	 */
	public static function getSession($namespace = NULL)
	{
		$handler = self::getService('Nette\Web\Session');
		return func_num_args() === 0 ? $handler : $handler->getNamespace($namespace);
	}



	/********************* global configuration ****************d*g**/



	/**
	 * Loads global configuration from file and process it.
	 * @param  string|Nette\Config\Config  file name or Config object
	 * @return \ArrayObject
	 */
	public static function loadConfig($file = NULL)
	{
		return self::$config = self::getConfigurator()->loadConfig($file);
	}



	/**
	 * Returns the global configuration.
	 * @param  string key
	 * @param  mixed  default value
	 * @return mixed
	 */
	public static function getConfig($key = NULL, $default = NULL)
	{
		if (func_num_args()) {
			return isset(self::$config[$key]) ? self::$config[$key] : $default;

		} else {
			return self::$config;
		}
	}

}
