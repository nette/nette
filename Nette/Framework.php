<?php

/**
 * This file is part of the Nette Framework (https://nette.org)
 * Copyright (c) 2004 David Grudl (https://davidgrudl.com)
 */

declare(strict_types=1);

namespace Nette;


/**
 * Framework version.
 */
class Framework
{
	public const
		NAME = 'Nette Framework',
		VERSION = '3.0',
		VERSION_ID = 30000,
		REVISION = '';
}
