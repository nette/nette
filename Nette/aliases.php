<?php

defined('NETTE_ALIASES') && empty(NETTE_ALIASES) || array_walk(new ArrayIterator(array(
	'Nette\Config\Configurator' => 'Nette\Configurator',
	'Nette\Config\CompilerExtension' => 'Nette\DI\CompilerExtension',
	'Nette\Diagnostics\Bar' => 'Tracy\Bar',
	'Nette\Diagnostics\BlueScreen' => 'Tracy\BlueScreen',
	'Nette\Diagnostics\Dumper' => 'Tracy\Dumper',
	'Nette\Diagnostics\FireLogger' => 'Tracy\FireLogger',
	'Nette\Diagnostics\Logger' => 'Tracy\Logger',
	'Nette\Latte\ParseException' => 'Latte\CompileException',
	'Nette\Latte\CompileException' => 'Latte\CompileException',
	'Nette\Latte\Compiler' => 'Latte\Compiler',
	'Nette\Latte\IMacro' => 'Latte\IMacro',
	'Nette\Latte\Parser' => 'Latte\Parser',
	'Nette\Latte\Macros\CoreMacros' => 'Latte\Macros\CoreMacros',
	'Nette\Latte\Macros\MacroSet' => 'Latte\Macros\MacroSet',
	'Nette\ArrayHash' => 'Nette\Utils\ArrayHash',
	'Nette\ArrayList' => 'Nette\Utils\ArrayList',
	'Nette\DateTime' => 'Nette\Utils\DateTime',
	'Nette\Image' => 'Nette\Utils\Image',
	'Nette\ObjectMixin' => 'Nette\Utils\ObjectMixin',
)), 'class_alias');
