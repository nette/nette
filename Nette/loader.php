<?php

/**
 * Nette Framework (version 2.1-dev released on $WCDATE$, http://nette.org)
 *
 * Copyright (c) 2004, 2013 David Grudl (http://davidgrudl.com)
 *
 * For the full copyright and license information, please view
 * the file license.txt that was distributed with this source code.
 */



/**
 * Check and reset PHP configuration.
 */
if (PHP_VERSION_ID < 50301) {
	throw new Exception('Nette Framework requires PHP 5.3.1 or newer.');
}

@header('Content-Type: text/html; charset=utf-8'); // @ - headers may be sent



/**
 * Load and configure Nette Framework.
 */
define('NETTE', TRUE);
define('NETTE_DIR', __DIR__);
define('NETTE_VERSION_ID', 20100); // v2.1.0



require_once __DIR__ . '/common/exceptions.php';
require_once __DIR__ . '/common/Object.php';
require_once __DIR__ . '/Utils/LimitedScope.php';
require_once __DIR__ . '/Loaders/AutoLoader.php';
require_once __DIR__ . '/Loaders/NetteLoader.php';


Nette\Loaders\NetteLoader::getInstance()->register();

require_once __DIR__ . '/Diagnostics/shortcuts.php';

Nette\Utils\SafeStream::register();
class_alias('Nette\Configurator', 'Nette\Config\Configurator');



/**
 * Nette\Callback factory.
 * @param  mixed   class, object, callable
 * @param  string  method
 * @return Nette\Callback
 */
function callback($callback, $m = NULL)
{
	return new Nette\Callback($callback, $m);
}
