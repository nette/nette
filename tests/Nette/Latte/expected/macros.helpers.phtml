<?php list($_l, $_g) = Nette\Latte\Macros\CoreMacros::initRuntime($template, 'xxx')
;
// snippets support
if (%A%}

//
// main template
//
?>
<ul>
	<li>lower: <?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->lower($hello), ENT_NOQUOTES) ?></li>
	<li>upper: <?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->upper($hello), ENT_NOQUOTES) ?></li>
	<li>lower & capitalize: <?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->capitalize($template->lower($hello)), ENT_NOQUOTES) ?></li>
	<li>nl2br: <?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->nl2br($hello), ENT_NOQUOTES) ?></li>
	<li>truncate: <?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->lower($template->truncate($hello, "10 ")), ENT_NOQUOTES) ?></li>
	<li>date: <?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->date($date, '%d.%m.%Y %H:%M:%S'), ENT_NOQUOTES) ?></li>
	<li>translated: <?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->truncate($template->translate($hello), 3), ENT_NOQUOTES) ?></li>
	<li>spaces: <?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->types($hello , '' , "" , "$hello" ), ENT_NOQUOTES) ?></li>
</ul>

<ul>
	<li><?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->h2($template->h1($hello)), ENT_NOQUOTES) ?></li>
	<li><?php echo $template->h2($template->h1($hello)) ?></li>
	<li><?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->h1($template->h2($hello)), ENT_NOQUOTES) ?></li>
	<li><?php echo $template->h1($template->h2($hello)) ?></li>
</ul>

<ul>
	<li><?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->types($hello*0, 0, 0.0, "0"), ENT_NOQUOTES) ?></li>
	<li><?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->types($hello*1, 1, "1"), ENT_NOQUOTES) ?></li>
	<li><?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->types($hello, true, null, false), ENT_NOQUOTES) ?></li>
	<li><?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->types($hello, TRUE, NULL, FALSE), ENT_NOQUOTES) ?></li>
	<li><?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->types($hello, '', "", "$hello"), ENT_NOQUOTES) ?></li>
</ul>



<?php ob_start() ?>  <a   href="#"
> test</a

>
AÅ A

<script>
// comment
alert();
</script>
<?php echo $template->strip(ob_get_clean()) ?>



<?php ob_start() ?>
<ul>
	<li>lower: <?php echo Nette\Templating\DefaultHelpers::escapeHtml($template->lower($hello), ENT_NOQUOTES) ?></li>
</ul>
<?php $capture = ob_get_clean() ?>

<p>
Captured: <?php echo Nette\Templating\DefaultHelpers::escapeHtml($capture, ENT_NOQUOTES) ?>

</p>


<?php ob_start() ?>
Hello
<?php $capture = $template->upper(ob_get_clean()) ?>

<p>
Captured with modifier: <?php echo Nette\Templating\DefaultHelpers::escapeHtml($capture, ENT_NOQUOTES) ?>

</p>



<p>
Nested blocks: <?php ob_start() ?> Outer   <?php ob_start() ?> Inner Block <?php echo $template->upper(ob_get_clean()) ?>
  Block <?php echo $template->truncate(ob_get_clean(), 20) ?>

</p>
