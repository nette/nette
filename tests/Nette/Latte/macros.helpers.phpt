<?php

/**
 * Test: Nette\Latte\Engine: helpers test.
 *
 * @author     David Grudl
 * @package    Nette\Latte
 * @subpackage UnitTests
 * @keepTrailingSpaces
 */

use Nette\Latte,
	Nette\Templating\FileTemplate;



require __DIR__ . '/../bootstrap.php';

require __DIR__ . '/Template.inc';



class MyHelper
{
	protected $count = 0;

	public function invoke($s)
	{
		$this->count++;
		return strtolower($s) . " ($this->count times)";
	}

}

function types()
{
	foreach (func_get_args() as $arg) $res[] = gettype($arg);
	return implode(', ', $res);
}



$template = new FileTemplate(__DIR__ . '/templates/helpers.latte');
$template->registerFilter(new Latte\Engine);
$template->registerHelper('nl2br', 'nl2br');
$template->registerHelper('h1', array(new MyHelper, 'invoke'));
$template->registerHelper('h2', 'strtoupper');
$template->registerHelper('translate', 'strrev');
$template->registerHelper('types', 'types');
$template->registerHelperLoader('Nette\Templating\DefaultHelpers::loader');

$template->hello = 'Hello World';
$template->date = strtotime('2008-01-02');

$path = __DIR__ . '/expected/' . basename(__FILE__, '.phpt');
Assert::match(file_get_contents("$path.phtml"), codefix($template->compile()));
Assert::match(file_get_contents("$path.html"), $template->__toString(TRUE));
