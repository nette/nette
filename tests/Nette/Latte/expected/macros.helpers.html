<ul>
	<li>lower: hello world</li>
	<li>upper: HELLO WORLD</li>
	<li>lower & capitalize: Hello World</li>
	<li>nl2br: Hello World</li>
	<li>truncate: hello…</li>
	<li>date: 02.01.2008 00:00:00</li>
	<li>translated: dl…</li>
	<li>Translated HTML: joha</li>
	<li>Translated HTML: joha</li>
	<li>Translated HTML: johajohajohajoha|joha</li>
	<li>spaces: string, string, string, string</li>
</ul>

<ul>
	<li>HELLO WORLD (1 TIMES)</li>
	<li>HELLO WORLD (2 TIMES)</li>
	<li>hello world (3 times)</li>
	<li>hello world (4 times)</li>
</ul>

<ul>
	<li>integer, integer, double, string</li>
	<li>integer, integer, string</li>
	<li>string, boolean, NULL, boolean</li>
	<li>string, boolean, NULL, boolean</li>
	<li>string, string, string, string</li>
</ul>



<a href="#" > test</a > AŠA<script>
// comment
alert();
</script>




<p>
Captured: &lt;ul&gt;
	&lt;li&gt;lower: hello world&lt;/li&gt;
&lt;/ul&gt;

</p>



<p>
Captured with modifier: HELLO

</p>



<p>
Nested blocks:  Outer    INNER…
</p>
